### DISCUSSION TITLE
Multiple OSC [netreceive]rs (or, sharing data between instances?)

### DISCUSSION THREAD

**Discussion by jamshark70** [plugdata-team/plugdata#1497] [Ideas]
For an electronic ensemble course, one idea was to allow students to use their DAW (for comfort). So they would have multiple tracks, of which only one or two would be playing at a time. We also wanted to get control data from mobile devices (OSC over WLAN) and route to VST parameters.

Multiple instances of [netreceive] do not cooperate. To test this, I set up a Reaper project with two tracks, each with its own instance of plugdata, running the same patch where [netreceive -u -b (port)] goes --> [oscparse] --> [route] etc. When I send OSC to the right port, the second track's plugdata responds to the message, but the first one does nothing.

Mobile device OSC apps usually can't send to multiple ports -- so it's necessary to have multiple [netreceive]s with the same port -- but this is usually not allowed ("netreceive: listen failed: Address already in use (98)").

I also thought to work around this in the Max4Live way, by exploiting the shared namespace for [send] / [receive] pairs. I was pretty sure this wouldn't work -- I'd have been very surprised if multiple libpds shared namespaces -- and indeed it didn't. (This test was: Track 1 has the [netreceive], and uses [send cutoff] for the control value. Track 2 omits [netreceive] but has [r cutoff]. Track 1 responds; track 2 does not.)

For this use case, then, it seems like my best option would be to have a standalone Max or Pd process that sends MIDI CCs to a loopback interface. Probably OK in the short term (so, no expectation of any near-term work on this) but it does make me wonder -- where there are use cases for multiple plug-in instances having access to the same external control stream, what would be a good way for plugdata to support that?

**Comment by jamshark70** [comment]
> Mobile device OSC apps usually can't send to multiple ports -- so it's necessary to have multiple [netreceive]s with the same port -- but this is usually not allowed ("netreceive: listen failed: Address already in use (98)").

FWIW there's one update: now, instead of posting an error, all instances (except for the one that actually responds) fail silently.

I'd like to refer also to this discussion on the PD forum where a user was trying to configure multiple receivers within the same DAW project: https://forum.pdpatchrepo.info/topic/15928/issue-with-netsend-receive

Here, the initial behavior that was reported is a bit different from what I reported. He says that all instances received something, but exactly the same data, even though they were supposed to be separate messages with separate arguments. I haven't studied the thread carefully, so I'm not entirely clear what's going on. **BUT** I mention this because, if use of plugdata as a plug-in increases, there will be more and more such cases: not only about net receipt, but also shared arrays, shared send/receive locations, etc.

I understand why nothing is shared currently: start up a pdlib and it's its own universe. OK. FWIW, Max4Live shares all identifiers across an entire project, which opens up some possibilities that are difficult/impossible with plugdata currently. (I usually prefer Pd's design decisions over Max's, but in this case, I think Ableton+Cycling74 got it right where plugdata is kinda casually accepting a limitation of pd as a library, and avoiding making design decisions related to that.)

**Reply by dromer** [reply]
It's kind of unfair to compare with Max4Live since that is not running as a plugin but built into the DAW.
You are not sharing anything between different contexts, since everything lives in the same process already.

**Reply by jamshark70** [reply]
> It's kind of unfair to compare with Max4Live since that is not running as a plugin but built into the DAW.

That is true, but this distinction doesn't invalidate the use cases.

**Reply by dromer** [reply]
But you are blaming "design decisions". However the facts of life are that we can only run plugdata as a plugin, and not build into the program directly.

**Reply by jamshark70** [reply]
> But you are blaming "design decisions"

OK, let me try another way to say it.

Typical PD vanilla usage would have one PD instance running on a machine at a time. (AFAICS it's impossible to run multiple PD processes off the same binary -- `./pd-gui` the first time launches the app; the second time, in a different terminal, switches to the process that's already running, but doesn't open a second process.)

So it's reasonable for PD vanilla [netreceive] to operate on the assumption that there will be only one receiver for a given port number, because you can design your patch to receive messages in one place and farm the messages out to wherever they're needed (even across patch boundaries with [s] and [r]).

Pd-as-a-plugin opens up the possibility of multiple instances on the same machine (even within the same process), which was basically never an option before. That is, *usage requirements are not standing still*, and my suggestion is that the software's behavior should adapt to a new use case that didn't exist until plugdata.

Now, it *is* a fair position, that plugdata may not wish to override PD core behavior -- i.e., that changes in netreceive should made in core and imported into plugdata from there. I opened the discussion here because I ran into the problem when using plugdata, but sure, the solution could be somewhere else. There's a risk here of mutual passing the buck, if Miller says "we're not officially supporting multiple simultaneous instances of PD vanilla, so I don't care about this issue" and plugdata says "we can't do anything that Miller hasn't approved." (Who knows, he might approve it -- I'll file a request about that in a little while.)

I do think it's beneficial for the plugdata team to acknowledge that there are new use cases and to consider whether there might be some way to handle them.

**Reply by dromer** [reply]
I can easily open multiple pd instances, so this assessment does not hold up.
With PD there hardly is any "typical" usage. Everybody uses it differently :)

**Reply by dromer** [reply]
You can do this just fine with PD on a single system:
<img width="1017" height="817" alt="2026-09-25_09-03" src="https://github.com/user-attachments/assets/5efdcbdf-d5f2-484c-b257-5aaa2189ebc8" />

**Reply by dromer** [reply]
And here with two separate plugdata vst3 instances:
<img width="1344" height="682" alt="2026-09-25_09-07" src="https://github.com/user-attachments/assets/32e1c98a-4447-4dbc-8161-c3ea8b7105f2" />

**Reply by dromer** [reply]
But as you said you cannot have multiple listeners on the same port.

**Comment by jamshark70** [comment]
> I can easily open multiple pd instances, so this assessment does not hold up.

I get the peculiar feeling at this point that the discussion has shifted focus away from evaluating usage requirements and toward debating the validity of claims tangential to the stated issue.

> But as you said you cannot have multiple listeners on the same port.

At this point, I'm just going to close this discussion and take the issue over to the main PD repository.

I followed up on this discussion a couple of days ago because a user other than myself brought up the same question on the patch repo forum. It may not be a common request, but somebody else at some point in the future will run into the same limitation. I'm looking for a way forward from here. As noted, a better first step is probably to see what's up with netreceive. Incidentally Max *does* allow multiple [udpreceive] objects on the same port; this supports the argument that [netreceive] may be unnecessarily strict when it clobbers prior connections on that port. But let's see what the devs say.

**Reply by dromer** [reply]
Ok, I now read your post more carefully. Sorry about that.

With TCP you can only receive once, but UDP should indeed work.
For UDP you indeed should be able to set `-u` flag, but I now tested this in plugdata and it's only receiving on one instance (trying standalone right now).

On paper the UDP version should work, I think. Or is it multiple senders to a single receiver?
Now I'm confused -_-

I'll let someone else chip in .. sorry for the extra noise.

**Reply by dromer** [reply]
I see the same issue in pd vanilla. With UDP only the first instantiated listener is able to receive messages, but given multicast you'd expect multiple to be able to access this port in UDP mode.

So indeed the best thing is to relay this to puredata repo and see what the higher ups have to say ;)

**Reply by jamshark70** [reply]
> With UDP only the first instantiated listener is able to receive messages

For me, it's the last-created [netreceive], but the substance here is correct: with multiple [netreceive] objects listening to the same port, only one of them ever fires.

Here, if both of them fire, you would hear an A and an F#. I hear only the F#.

<img width="366" height="380" alt="pd-netreceive-2" src="https://github.com/user-attachments/assets/ef0cf5b2-8f91-4854-87c1-10382356e410" />

```
#N canvas 1428 131 1070 905 12;
#X obj 57 54 netreceive -u -b 9999;
#X obj 57 79 oscparse;
#X obj 57 104 list trim;
#X obj 57 129 route play;
#X msg 57 194 1 100 \, 0 3000 100;
#X obj 57 219 vline~;
#X obj 57 291 *~;
#X obj 75 263 osc~ 440;
#X obj 57 370 dac~;
#X obj 57 338 *~ 0.2;
#X obj 226 54 netreceive -u -b 9999;
#X obj 226 79 oscparse;
#X obj 226 104 list trim;
#X obj 226 129 route play;
#X msg 226 194 1 100 \, 0 3000 100;
#X obj 226 219 vline~;
#X obj 226 291 *~;
#X obj 244 263 osc~ 366;
#X connect 0 0 1 0;
#X connect 1 0 2 0;
#X connect 2 0 3 0;
#X connect 3 0 4 0;
#X connect 4 0 5 0;
#X connect 5 0 6 0;
#X connect 6 0 9 0;
#X connect 7 0 6 1;
#X connect 9 0 8 0;
#X connect 9 0 8 1;
#X connect 10 0 11 0;
#X connect 11 0 12 0;
#X connect 12 0 13 0;
#X connect 13 0 14 0;
#X connect 14 0 15 0;
#X connect 15 0 16 0;
#X connect 16 0 9 0;
#X connect 17 0 16 1;
```

> So indeed the best thing is to relay this to puredata repo and see what the higher ups have to say.

https://github.com/pure-data/pure-data/issues/2969#issuecomment-5832573471 ... which, I'll give that another day or so and then open a new issue if no reply.

Also I do have a feeling that the pd-as-plugin situation may not be possible, since (IIUC) plugins really shouldn't be sharing data outside of input and output channels. But, the question has come up.
